### PR TITLE
[MENFORCER-361] optionally normalize line separators prior to checksum calculation

### PR DESCRIPTION
**Pull request by kwin**
Normalizing line separators is useful to properly support multi-platform development where text-based files are supposed to differ in their line separators (which should not have an impact on the checksum).

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MENFORCER) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MENFORCER-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MENFORCER-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by kwin** [comment]
Can someone have a look at this PR? Is there anything missing?

**Comment by michael-o** [comment]
Can you describe the usecase for this?

**Comment by kwin** [comment]
@michael-o Hopefully this is clear from https://issues.apache.org/jira/browse/MENFORCER-361. I want to use platform independent checksums for text files.

**Comment by michael-o** [comment]
I see, do you already have a usecase for this?

**Comment by kwin** [comment]
Yes, look at https://github.com/adobe/aem-project-archetype/pull/480 where this is currently leveraged with two different profiles: One for Windows and one for Linux/Mac OS.

**Comment by gmshake** [comment]
Hi, @kwin 

The current enforcer `RequireFileChecksum` checks `file` checksum. For `text` files, line separator varies on different platform, OS X, Windows, Linux etc. For `binary` files, it is confusing to normalize the line separator. So, I'd propose we introduce a new `RequireTextFileChecksum` enforcer to make it clear.

**Comment by kwin** [comment]
> The current enforcer RequireFileChecksum checks file checksum

Agree, but that includes both binary and text files. There is no hint in https://maven.apache.org/enforcer/enforcer-rules/requireFileChecksum.html that this should only be used for binaries.

> So, I'd propose we introduce a new RequireTextFileChecksum enforcer to make it clear.

I am fine with that as well, but honestly I don't see any benefit here. As both need configuration per file, one can just as well add another configuration flag (`normalizeLineSeparatorTo`). Why do you prefer two different rule names here?

**Comment by gmshake** [comment]
@kwin

> Agree, but that includes both binary and text files. There is no hint in https://maven.apache.org/enforcer/enforcer-rules/requireFileChecksum.html that this should only be used for binaries.

We commonly calculate files's checksum regardless the content type, i.e.  we always treat them as `binary`. Since your usecase is special, normalizing line separator before calculating file's checksum, it's better not to confuse end user with the common usecase. 

Anyway, both works :)

**Comment by michael-o** [comment]
I think this can be drastically simplified by saying:
* we have a `requireTextFileChecksum`
* Convert into a canonical text form (Unix style endings)

No further configuration will be required. This should be idiot-proof.

WDYT?

**Comment by kwin** [comment]
Ok, will modify the PR accordingly. Notice though that the character encoding configuration is mandatory though (but I will use `${project.build.sourceEncoding}` by default)

**Comment by michael-o** [comment]
Right, even on UTF-16 it will be four bytes for `\r\n`.

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * @author Konrad Windszus
+ * @see RequireFileChecksum
+ */
+public class RequireTextFileChecksum
+    extends RequireFileChecksum
+{
+    private NormalizeLineSeparatorReader.LineSeparator normalizeLineSeparatorTo = LineSeparator.UNIX;
+
+    private Charset charsetEncoding;
```
Comment (michael-o) [inline enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java:48]: make it just `encoding` akin to the setter

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    public void setNormalizeLineSeparatorTo( NormalizeLineSeparatorReader.LineSeparator normalizeLineSeparatorTo )
+    {
+        this.normalizeLineSeparatorTo = normalizeLineSeparatorTo;
+    }
+
+    public void setEncoding( String fileCharset )
```
Comment (michael-o) [inline enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java:55]: Arg too

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        // set defaults
+        if ( charsetEncoding == null )
+        {
+            // https://maven.apache.org/plugins/maven-resources-plugin/examples/encoding.html
+            try
+            {
+                String encoding = (String) helper.evaluate( "${project.build.sourceEncoding}" );
+                if ( StringUtils.isBlank( encoding ) )
+                {
+                    throw new EnforcerRuleException( "No encoding set for '${project.build.sourceEncoding}'" );
```
Comment (michael-o) [inline enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java:73]: This is wrong. It should be exactly like here: https://maven.apache.org/plugins/maven-resources-plugin/xref/org/apache/maven/plugins/resources/ResourcesMojo.html#L313

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+        super.execute( helper );
+    }
+
+    @Override
+    protected String calculateChecksum()
+        throws EnforcerRuleException
+    {
+        try ( FileInputStream fileInputStream = new FileInputStream( this.file );
```
Comment (michael-o) [inline enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java:90]: Why so complex? What about `Files.newInputStream()` or `Files.newBufferedReader()`?

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/utils/NormalizeLineSeparatorReader.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+{
+
+    private static final int EOL = -1;
+
+    /** 
+     * Type representing either Unix or Windows line separators
+     */
+    public enum LineSeparator
+    {
+        WINDOWS( "\r\n", null ), UNIX( "\n", '\r' );
```
Comment (michael-o) [inline enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/utils/NormalizeLineSeparatorReader.java:40]: UNIX does not have `\r` as any line separator. Why is it there?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/utils/NormalizeLineSeparatorReader.java (hunk 5 above)
Comment (michael-o) [reply enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/utils/NormalizeLineSeparatorReader.java:40]: Ah thanks!

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java (hunk 4 above)
Comment (kwin) [reply enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java:90]: Good point, gonna simplify, left-over from refactoring.

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java (hunk 1 above)
Comment (kwin) [reply enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java]: Fixed in https://github.com/apache/maven-enforcer/pull/79/commits/731853114214d25decf8c6a1eab726cd82783348.

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java (hunk 2 above)
Comment (kwin) [reply enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java]: https://github.com/apache/maven-enforcer/pull/79/commits/731853114214d25decf8c6a1eab726cd82783348

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java (hunk 4 above)
Comment (kwin) [reply enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java]: https://github.com/apache/maven-enforcer/pull/79/commits/731853114214d25decf8c6a1eab726cd82783348

File: enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java (hunk 3 above)
Comment (kwin) [reply enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequireTextFileChecksum.java]: https://github.com/apache/maven-enforcer/pull/79/commits/731853114214d25decf8c6a1eab726cd82783348

**Review by michael-o — CHANGES_REQUESTED** [review]
Almost merge ready

**Comment by michael-o** [comment]
Thanks, I will test tomorrow and merge. Ping me at 16:00 UTC if I forget this.

**Comment by kwin** [comment]
The test failure in Windows, Java11 with GitHub Actions at https://github.com/apache/maven-enforcer/pull/79/checks?check_run_id=1260193340 doesn't seem related to my change (`org.apache.maven.plugins.enforcer.TestEvaluateBeanshell.testRuleCanExecuteMultipleThreads()`).
@michael-o Let me know if I should squash the commits.

**Comment by michael-o** [comment]
I will have a look tomorrow at it.

**Comment by kwin** [comment]
@michael-o Thanks for merging.

**Pull request closed without merging**